src/catalog-debug/src/main.rs

```diff
@@ -0,0 +1,519 @@
+    /// or error message. Exits with 0 if the upgrade would succeed, otherwise
+    /// non-zero. Can be used on a running environmentd. Operates without
+    /// interfering with it or committing any data to that stash.
+    UpgradeCheck {
+        /// Map of cluster name to resource specification. Check the README for latest values.
+        cluster_replica_sizes: Option<String>,
+    },
+}
+
+#[tokio::main]
+async fn main() {
+    let args = cli::parse_args(CliConfig {
+        env_prefix: Some("MZ_CATALOG_DEBUG_"),
+        enable_version_flag: true,
+    });
+    if let Err(err) = run(args).await {
+        eprintln!("catalog: fatal: {}", err.display_with_causes());
+        process::exit(1);
+    }
+}
+
+async fn run(args: Args) -> Result<(), anyhow::Error> {
+    let metrics_registry = MetricsRegistry::new();
+    let openable_state: Box<dyn OpenableDurableCatalogState> = match args.store {
+        CatalogKind::Stash => {
+            let postgres_url = args.postgres_url.expect("required for stash");
+            let tls =
+                mz_tls_util::make_tls(&tokio_postgres::config::Config::from_str(&postgres_url)?)?;
+            let factory = StashFactory::new(&metrics_registry);
+            let stash_config = StashConfig {
+                stash_factory: factory,
+                stash_url: postgres_url,
+                schema: None,
+                tls,
+            };
+            Box::new(stash_backed_catalog_state(stash_config))
+        }
+        CatalogKind::Persist => {
+            let persist_config =
+                PersistConfig::new(&mz_environmentd::BUILD_INFO, SYSTEM_TIME.clone());
+            let persist_pubsub_server =
+                PersistGrpcPubSubServer::new(&persist_config, &metrics_registry);
+            let persist_pubsub_client = persist_pubsub_server.new_same_process_connection();
+
+            let persist_clients =
+                PersistClientCache::new(persist_config, &metrics_registry, |_, metrics| {
+                    let sender: Arc<dyn PubSubSender> = Arc::new(
+                        MetricsSameProcessPubSubSender::new(persist_pubsub_client.sender, metrics),
+                    );
+                    PubSubClientConnection::new(sender, persist_pubsub_client.receiver)
+                });
+            let persist_location = PersistLocation {
+                blob_uri: args
+                    .persist_blob_url
+                    .expect("required for persist")
+                    .to_string(),
+                consensus_uri: args
+                    .persist_consensus_url
+                    .expect("required for persist")
+                    .to_string(),
+            };
+            let persist_client = persist_clients.open(persist_location).await?;
+            let organization_id = args.organization_id.expect("required for persist");
+            Box::new(persist_backed_catalog_state(persist_client, organization_id).await)
```

Comment:
@danhhz can you take a look at this and make sure it makes sense? I pretty much copied it from `environmentd`. For the build info do I need `&mz_environmentd::BUILD_INFO` or can I just use the build info from this crate?

Reply:
This `CatalogKind::Persist` looks good modulo the pubsub comment I left above. Persist can read versions from the past, but if you're going to do any persist writes, the `BUILD_INFO` must be one of the ones with a version kept in lock-step with the others (there's a list in `bin/bump-version`). Otherwise you might brick the env (by writing a version of persist state in the future, that the env will refuse to use and then panic on)

Reply:
Ok cool, I've added this crate to `bin/bump-version` as part of this PR, so I'll just use this crate's build version.